api/v1beta1/statemanagementprovider_types.go

```diff
@@ -0,0 +1,204 @@
+	// StateManagementProviderFailedProvisionerEvent indicates the event for the provisioners failure
+	StateManagementProviderFailedProvisionerEvent = "FailedToEnsureProvisioner"
+	// StateManagementProviderSuccessProvisionerEvent indicates the event for the provisioners success
+	StateManagementProviderSuccessProvisionerEvent = "SuccessfullyEnsuredProvisioner"
+	// StateManagementProviderFailedProvisionerCRDsEvent indicates the event for the ProvisionerCRDs failure
+	StateManagementProviderFailedProvisionerCRDsEvent = "FailedToEnsureProvisionerCRDs"
+	// StateManagementProviderSuccessProvisionerCRDsEvent indicates the event for the ProvisionerCRDs success
+	StateManagementProviderSuccessProvisionerCRDsEvent = "SuccessfullyEnsuredProvisionerCRDs"
+)
+
+// StateManagementProviderSpec defines the desired state of StateManagementProvider
+type StateManagementProviderSpec struct {
+	// +kubebuilder:default=false
+
+	// Suspend suspends the StateManagementProvider. Suspending a StateManagementProvider
+	// will prevent the adapter from reconciling any resources.
+	Suspend bool `json:"suspend"`
+
+	// Adapter is an operator with translates the k0rdent API objects into provider-specific API objects.
+	// It is represented as a reference to operator object
+	Adapter ResourceReference `json:"adapter"`
+
+	// Provisioner is a set of resources required for the provider to operate. These resources
+	// reconcile provider-specific API objects. It is represented as a list of references to
+	// provider's objects
+	Provisioner []ResourceReference `json:"provisioner"`
+
+	// ProvisionerCRDs is a set of references to provider-specific CustomResourceDefinition objects,
+	// which are required for the provider to operate.
+	ProvisionerCRDs []ProvisionerCRD `json:"provisionerCRDs"`
+}
+
+// ResourceReference is a cross-namespace reference to a resource
+type ResourceReference struct {
+	// APIVersion is the API version of the resource
+	APIVersion string `json:"apiVersion"`
+
+	// Kind is the kind of the resource
+	Kind string `json:"kind"`
+
+	// Name is the name of the resource
+	Name string `json:"name"`
+
+	// Namespace is the namespace of the resource
+	Namespace string `json:"namespace"`
+
+	// ReadinessRule is a CEL expression that evaluates to true when the resource is ready
+	ReadinessRule string `json:"readinessRule"`
+}
+
+// ProvisionerCRD is a GVRs for a custom resource reconciled by provisioners
+type ProvisionerCRD struct {
+	// Group is the API group of the resources
+	Group string `json:"group"`
+
+	// Version is the API version of the resources
+	Version string `json:"version"`
+
+	// Resources is the list of resources under given APIVersion
+	Resources []string `json:"resources"`
+}
+
+// StateManagementProviderStatus defines the observed state of StateManagementProvider
+type StateManagementProviderStatus struct {
+	// Conditions is a list of conditions for the state management provider
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
```

Comment:
shouldn't we add the following as recommended by the type definition?

```yaml
// +patchMergeKey=type
// +patchStrategy=merge
// +listType=map
// +listMapKey=type
```

there are other resources that also miss these keys (that is just a notice that the current code base also misses it)